enhancement/pca estimate of no of signifcant components

pquinn-dls opened the pull request:
I've added a method to estimate the no of signficant components in PCA - enhancement on the current threshold method.
This performs a simple measurements of where the elbow position is and stores it. 
This calculation is performed in the decomposition method and when using plot_explained_variance_ratio  setting vline=True will plot the 
vertical line for this estimated position. 

![screeplot3](https://user-images.githubusercontent.com/30322591/56284006-2b453080-610b-11e9-93c1-9d0639636561.png)

A property no_significant_components has also been added to LearningResults. Storing the value is to help with automated scripts and later cluster analysis commits.
Small change also added to user guide

pquinn-dls committed:
pca estimate of no of signifcant components

jat255 commented:
@pquinn-dls thanks for the contribution! At first glance, this looks pretty good, and very useful! Is there a reference that you could include in the documentation to using this method? It looks like a fairly obvious method once I read the comments in the code, but I hadn't seen it used before. 

A couple of requests:
- [x] Can you please add a few test cases and ensure that they pass? You can look in the `test` folder for examples using the `pytest` framework. I would create a few dummy signals with a known number of significant components, and then ensure that the method returns the expected result.
- [x] If it's not too much trouble, it might be nice to have the "triangle" visualized at least in the documentation to make it clear how the point is determined. Here's a horrible quick example of what I mean using a screenshot tool:
   - ![image](https://user-images.githubusercontent.com/1278301/56299897-ecab7800-6102-11e9-90c7-dd3426cd0c65.png)
- [x] It seems to me that in your example, the 4th component is the furthest from the "triangle line", but the `vline` appears to be drawn through the 5th component. Is this an error, or am I just not understanding how the method works correctly?

pquinn-dls committed:
Added test for elbow position method

pquinn-dls committed:
image of elbow method and updated user guide

pquinn-dls committed:
fixed no signifcant components vs index of elbow

pquinn-dls committed:
d reference for knee method and updated doc.

pquinn-dls commented:
Hi,

Yes you're correct the elbow is the 4th but the no of significant pca components was set to elbow + 1
I realised that I was plotting the no of components - not the index so I've adjusted that. 

I've added a reference, some additional docs and an example image of the method.

Ville Satopää, Jeannie Albrecht, David Irwin, Barath Raghavan. Finding a "Kneedle" in a Haystack: Detecting Knee Points in System Behavior. 31st International Conference on Distributed Computing Systems Workshops, pp. 166-171, Minneapolis, Minnesota, USA, June 2011.

I've also added a test but I'm not sure if I've done this correctly - need to read a bit more on this!

jat255 commented on doc/user_guide/mva.rst:
```diff
@@ -104,8 +104,16 @@ the total variance specified, and the components above this value will be
 for ``threshold``, the line will be drawn at the specified component (see
-below).  The "elbow" position can be auto-estimated and a vertical line drawn to
-represent the no of signicant pca components by specifying "vline" as True. 
+below).  The no of significant components can be estimated and a vertical line drawn to
```
`no` --> `number`

pquinn-dls replied:
changed

jat255 commented on doc/user_guide/mva.rst:
```diff
@@ -104,8 +104,16 @@ the total variance specified, and the components above this value will be
-below).  The "elbow" position can be auto-estimated and a vertical line drawn to
-represent the no of signicant pca components by specifying "vline" as True. 
+below).  The no of significant components can be estimated and a vertical line drawn to
+represent this by specifying "vline" as True. In this case the elbow or knee is found 
```
"vline" --> \`\`vline\`\`

jat255 commented on doc/user_guide/mva.rst:
```diff
@@ -104,8 +104,16 @@ the total variance specified, and the components above this value will be
-below).  The "elbow" position can be auto-estimated and a vertical line drawn to
-represent the no of signicant pca components by specifying "vline" as True. 
+below).  The no of significant components can be estimated and a vertical line drawn to
+represent this by specifying "vline" as True. In this case the elbow or knee is found 
```
"In this case the elbow" --> "In this case, the elbow"

pquinn-dls replied:
fixed

jat255 commented on doc/user_guide/mva.rst:
```diff
@@ -104,8 +104,16 @@ the total variance specified, and the components above this value will be
+below).  The no of significant components can be estimated and a vertical line drawn to
+represent this by specifying "vline" as True. In this case the elbow or knee is found 
+in the variance plot by estimating the distance from each point in the variance plot 
+to a line joining the first and last points of the plot.  The no of significant PCA components 
```
"no" --> "number"

pquinn-dls replied:
fixed

pquinn-dls committed:
changes in manual no -> number
and fixed test to reflect change from number of components to index in elbow method

ericpre committed:
Fix tests.

ericpre committed:
Fix issue with storing number significant component and reset this value when starting decomposition (for algorithm not setting explained variance).

jat255 commented on doc/user_guide/mva.rst:
```diff
@@ -104,11 +104,11 @@ the total variance specified, and the components above this value will be
-below).  The no of significant components can be estimated and a vertical line drawn to
-represent this by specifying "vline" as True. In this case the elbow or knee is found 
+below).  The number of significant components can be estimated and a vertical line drawn to
+represent this by specifying ''vline'' as True. In this case, the elbow or knee is found 
```
Sorry, can you change this to use backticks (\`) instead of single quotes (like \`\`threshold\`\` above it)? That will cause the value to be formatted in monospace in the generated documentation.

jat255 commented:
Also, there were some merges into `RELEASE_next_minor`  yesterday. Can you please rebase your branch onto an updated copy of `RELEASE_next_minor`?

pquinn-dls committed:
fixed doc vline from '' to ``

Commit:
add web links to bibliography entries

Commit:
correct some formatting in user guide

Commit:
rename some variables to be more explanatory (no vs. num); PEP8 corrections; some more descriptive comments

pquinn-dls commented:
Rebased to RELEASE_next_minor and fixed the vline documentation.

ericpre committed:
Skip test when sklearn is not installed.

Commit:
add numpy intersphinx; correct some documentation and add links

jat255 commented:
I made some suggested changes in https://github.com/pquinn-dls/hyperspy/pull/2, and it looks like @ericpre did so as well in https://github.com/pquinn-dls/hyperspy/pull/1. Can you please review these and incorporate them onto your branch? After that I think this PR will be good to go.

Commit:
fix a mistakenly removed line from user guide when resolving conflicts

jat255 commented on hyperspy/learn/mva.py:
```diff
@@ -1222,13 +1248,54 @@ def undo_treatments(self):
+        x1 = 0
+        # loop 
+        distance = np.zeros((maxpoints))
+        for i in range(maxpoints):
```
I'll add here that it would be better to do this with a vector calculation (https://stackoverflow.com/a/46470401/1435788), rather than looping through the array, but since it's such a small one-dimensional array, there'e probably no harm in leaving it as a loop.

pquinn-dls replied:
If it's ok I'll leave it as is for now as I think it's negligible computationally compared to the decomposition process in the majority of cases.

pquinn-dls replied:
I've merged and edited the changes and comments from @jat255  and @ericpre

pquinn-dls committed:
Merge pull request #1 from ericpre/pca_estimate_components

Fix tests for Pca estimate components

pquinn-dls committed:
Merge branch 'pca_estimate_components' into pca_estimate

pquinn-dls committed:
Merge pull request #2 from jat255/pca_estimate

Suggested changes

pquinn-dls committed:
clean up after merge

pquinn-dls committed:
modify vline explanation in docstrings

tjof2 commented on doc/user_guide/mva.rst:
```diff
@@ -104,7 +104,21 @@ the total variance specified, and the components above this value will be
+   :align:   center
+   :width:   500
+
+More details about the elbow or knee finding technique can be found in
```
I think more details are needed here rather than just referring back to the paper:

> In this case, the elbow or knee is found in the variance plot by estimating the distance from each point in the variance plot to a line joining the first and last points of the plot.

We should be clear about e.g. choosing between multiple candidates.

pquinn-dls replied:
Updated the description and commented on multiple candidates.

pquinn-dls committed:
updated explanation in documentation

pquinn-dls committed:
Typo, In the case of multiple

jat255 commented:
This looks pretty good to me, but will have to wait for @ericpre or @francisco-dlp to merge, since they have write access.

pquinn-dls commented:
Great, thanks for your help

ericpre reviewed (commented):
It looks good to me too. I have one comment: rename `num_significant_components` to `number_significant_components` to follow [PEP8](https://www.python.org/dev/peps/pep-0008/)? 
Even if this can be regarded as a matter of taste, we usually use full words as naming conventions and using `number` instead of `num` doesn't make it significantly longer. I am so used to it now that I was mislead by the original `no_significant_components`.

pquinn-dls committed:
Change num to number

pquinn-dls committed:
Changed num to number in tests

ericpre commented:
The failures are related to the releases of upstream libraries (numpy and mrcz).

ericpre commented:
Thanks @pquinn-dls.
